lightning/src/chain/channelmonitor.rs

```diff
@@ -1144,8 +1144,31 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 
 	/// Provides a payment_hash->payment_preimage mapping. Will be automatically pruned when all
 	/// commitment_tx_infos which contain the payment hash have been revoked.
-	pub(crate) fn provide_payment_preimage(&mut self, payment_hash: &PaymentHash, payment_preimage: &PaymentPreimage) {
+	pub(crate) fn provide_payment_preimage<B: Deref, F: Deref, L: Deref>(&mut self, payment_hash: &PaymentHash, payment_preimage: &PaymentPreimage, broadcaster: &B, fee_estimator: &F, logger: &L)
+	where B::Target: BroadcasterInterface,
+		    F::Target: FeeEstimator,
+		    L::Target: Logger,
+	{
 		self.payment_preimages.insert(payment_hash.clone(), payment_preimage.clone());
+		// If the channel is force closed, try to claim the output from this preimage
+		if self.lockdown_from_offchain || self.holder_tx_signed {
```

Comment:
I think this code doesn't cover the TODO left in `test_unordered_onchain_htlc_settlement` : "reverse the test with Bob's commitment reaching the chain instead of Alice's one".

Instead of relying on `lockdown_from_offchain`/`holder_tx_signed` to decide to broadcast a preimage tx, we should try :
* a) in the presence of counterparty commitment  onchain, broadcast a HTLC-Preimage transaction. Note you need to check both `current_counterparty_commitment_txid`/`prev_counterparty_commitment_txid` against `counterparty_commitment_txn_on_chain`.
* b) in the absence of counterparty commitment onchain, just try to generate a HTLC-Success for both `holder_commitment_txid`/`prev_holder_commitment_txid`. We will broadcast them but doesn't matter if we have not broadcast previously a commitment transaction. They won't neither propagate or confirm due to their orphaness.

I guess b) for now is okay even if in the future we should clean out `lockdown_from_offchain`/`holder_tx_signed` with some better flag also signaling a commitment has been seen onchain.

I think `claim_counterparty_htlc` is okay for now but I would lean towards reusing the same codepath that `block_connected` to avoid code duplication. FYI, avoiding duplication was one of the motivation of introducing `OnchainTxHandler`. Also `get_counterparty_htlc_output_claims_reqs` should be clean out after #642 (don't worry for the conflict, I'll handle it).

Reply:
Agree with most of your comment. Only thing:
> We will broadcast them but doesn't matter if we have not broadcast previously a commitment transaction. They won't neither propagate or confirm due to their orphaness.

Hmm, it feels weird to always broadcast two HTLC-success transactions on receipt of a preimage. I think we should only do this if `holder_tx_signed` is true, since if it hasn't been signed there's no way it's been seen on-chain. (Alternately we could replace `holder_tx_signed` with a map, `holder_commitment_tx_broadcasted<Txid, u64>` that mirrors `counterparty_commitment_txn_on_chain`).

Reply:
Lmk what you think of the updated solution.

Reply:
> I think we should only do this if holder_tx_signed is true, since if it hasn't been signed there's no way it's been seen on-chain. 

There is a caveat where this might not be true, in case of distributed watchtowers :
* Waldo broadcast the holder commitment with an incoming HTLC output
* The holder commitment confirms in block X
* Wally sees blocks X, does nothing, not preimage yet
* Wally receives a preimage from offchain, its `holder_tx_signed` might still be false

It's weird but make it easier to reason for this kind of scenario :)

Reply:
Ah OK I see what you mean. I switched it so now we broadcast the 2 preimage claiming transactions if `self.broadcasted_holder_revokable_script.is_some()`. I think this should work because it'll be non-`None` if we see any of our commitment txs onchain. Lmk what you think.

Reply:
Can you document the usage with the above scenario, at least mention it's done due to distributed watchtowers setup ? Current comment is saying "Then if a holder commitment transaction has been signed", it's a bit ambiguous as laid-out scenario is a case where commitment has been signed but not by _this_ local instance.